plugins/api/src/main/kotlin/PluginFactoryGenerator.kt

```diff
@@ -0,0 +1,352 @@
+        }
+
+        require(constructors.size == 1) {
+            "Plugin class $pluginClass must have exactly one constructor with a PluginDescriptor and a config " +
+                "argument."
+        }
+
+        return constructors.first()
+    }
+
+    /**
+     * Get the plugin options from the config class by mapping its properties to [PluginOption] instances.
+     */
+    @OptIn(KspExperimental::class)
+    private fun KSClassDeclaration.getPluginOptions(): List<PluginOption> {
+        require(Modifier.DATA in modifiers) {
+            "Config class $this must be a data class."
+        }
+
+        require(getConstructors().toList().size == 1) {
+            "Config class $this must have exactly one constructor."
+        }
+
+        val constructor = getConstructors().single()
+
+        return constructor.parameters.map { param ->
+            val paramType = param.type.resolve()
+            val paramTypeString = getQualifiedNameWithTypeArguments(paramType)
+            val paramName = param.name?.asString()
+
+            requireNotNull(paramName) {
+                "Config class constructor parameter has no name."
+            }
+
+            require(param.isVal) {
+                "Config class constructor parameter $paramName must be a val."
+            }
+
+            require(!param.hasDefault) {
+                "Config class constructor parameter $paramName must not have a default value. Default values must be " +
+                    "set via the @OrtPluginOption annotation."
+            }
+
+            val prop = getAllProperties().find { it.simpleName.asString() == paramName }
+
+            requireNotNull(prop) {
+                "Config class must have a property with the name $paramName."
+            }
+
+            val annotations = prop.getAnnotationsByType(OrtPluginOption::class).toList()
+
+            require(annotations.size <= 1) {
+                "Config class constructor parameter $paramName must have at most one @OrtPluginOption annotation."
+            }
+
+            val annotation = annotations.firstOrNull()
+
+            val type = when (paramTypeString) {
+                "kotlin.Boolean" -> PluginOptionType.BOOLEAN
+                "kotlin.Int" -> PluginOptionType.INTEGER
+                "kotlin.Long" -> PluginOptionType.LONG
+                "org.ossreviewtoolkit.plugins.api.Secret" -> PluginOptionType.SECRET
+                "kotlin.String" -> PluginOptionType.STRING
+                "kotlin.collections.List<kotlin.String>" -> PluginOptionType.STRING_LIST
+
+                else -> throw IllegalArgumentException(
+                    "Config class constructor parameter ${param.name?.asString()} has unsupported type " +
+                        "$paramTypeString."
+                )
+            }
+
+            val defaultValue = annotation?.defaultValue
+
+            PluginOption(
+                name = param.name?.asString().orEmpty(),
```

Comment:
Why not trim here as well, to be on the safe side?

Reply:
I think in Kotlin parameter names can actually have surrounding whitespace if you define them like:

```kotlin
 ` name`: String
```

In this case trimming would be wrong. The generated factory would still not compile, but I think this can be fixed if someone ever really needs it.

Reply:
I was referring to trimming the contents of the `name` string, just as you do for the `description`. Did we misunderstand each other?

Reply:
No, `param.name` is the name of the constructor argument which could start with whitespace in Kotlin because this is valid: `data class Test(val ` test`: String)`. Trimming the name would break it because it would not match the actual name anymore. So I'd rather keep the correct name here and fix it in the code generator if ever required (which I believe will never be the case, because who names parameters like that).